librz/core/cmp.c

```diff
@@ -0,0 +1,465 @@
+// SPDX-FileCopyrightText: 2021 Dhruv Maroo <dhruvmaru007@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_cmp.h>
+#include <rz_asm.h>
+#include <rz_list.h>
+
+/**
+ * \brief Compare memory at \p addr1 with the memory at \p addr2
+ *
+ * \param core Current RzCore instance
+ * \param addr1 address to read data from
+ * \param addr2 address to read data from
+ * \param len Number of bytes to compare
+ * \return RzCompareData* A pointer to RzCompareData comparison
+ */
+RZ_API RZ_OWN RzCompareData *rz_core_cmp_mem_mem(RzCore *core, ut64 addr1, ut64 addr2, ut32 len) {
+	rz_return_val_if_fail(core, NULL);
+
+	ut8 *buf1 = malloc(len * sizeof(ut8));
+	ut8 *buf2 = malloc(len * sizeof(ut8));
+	if (!buf1 || !buf2) {
+		goto error_goto;
+	}
+	if (rz_io_nread_at(core->io, addr1, buf1, len) == -1 || rz_io_nread_at(core->io, addr2, buf2, len) == -1) {
+		RZ_LOG_ERROR("Cannot read at provided addresses: 0x%" PFMT64x " 0x%" PFMT64x "\n", addr1, addr2);
+		goto error_goto;
+	}
+	RzCompareData *cmp = RZ_NEW0(RzCompareData);
+	if (!cmp) {
+		goto error_goto;
+	}
+	cmp->len = len;
+	cmp->data1 = buf1;
+	cmp->addr1 = addr1;
+	cmp->data2 = buf2;
+	cmp->addr2 = addr2;
+	cmp->same = rz_mem_eq(cmp->data1, cmp->data2, len);
+	return cmp;
+
+error_goto:
+	free(buf1);
+	free(buf2);
+	return NULL;
+}
+
+/**
+ * \brief Compare mem at \p addr with data \p data
+ *
+ * \param core Current RzCore instance
+ * \param addr address to read data from
+ * \param data Data to be compared
+ * \param len Number of bytes to compare
+ * \return RzCompareData* A pointer to RzCompareData comparison (the data1 corresponds to the data at addr (and addr1 = addr) and data2 is the str data (and addr2 = UT32_MAX))
+ */
+RZ_API RZ_OWN RzCompareData *rz_core_cmp_mem_data(RzCore *core, ut64 addr, RZ_NONNULL const ut8 *data, ut32 len) {
+	rz_return_val_if_fail(core && data, NULL);
+
+	ut8 *buf1 = malloc(len * sizeof(ut8));
+	if (!buf1) {
+		RZ_LOG_ERROR("Cannot read at address: 0x%" PFMT64x "\n", addr);
+		goto error_goto;
+	}
+	if (rz_io_nread_at(core->io, addr, buf1, len) == -1) {
```

Comment:
also here, the result should be used to make sure you really read all the data you want.

Reply:
Yes. I'll add the checks. That'll also make it easy for me to write the tests where the data read is less than the data expected (earlier, such test used to cause a segfault).